content/300-guides/050-database/870-using-prisma-with-mongodb.mdx

```diff
@@ -0,0 +1,351 @@
+
+</Admonition>
+
+These iterative tasks of updating the schema and the database can result in inconsistencies between your schema and the actual data in the database. Let’s look at one scenario where this can happen, and then examine several strategies for you and your team to consider for handling these inconsistencies.
+
+**Scenario**: you need to add a new field to your schema to accommodate a billing address that is separate from the currently stored shipping address. You currently have the following schema in your `schema.prisma` file:
+
+```prisma file=prisma/schema.prisma
+model Product {
+  id     String  @id @default(auto()) @map("_id") @db.ObjectId
+  name   String  @unique
+}
+
+model Order {
+  id              String   @id @default(auto()) @map("_id") @db.ObjectId
+  product         Product  @relation(fields: [productId], references: [id])
+  shippingAddress Address
+  productId       String   @db.ObjectId
+}
+
+type Address {
+  street String @id @default(auto()) @map("_id") @db.ObjectId
+  city   String
+  zip    String
+}
+```
+
+There are a number of strategies you could use for migrating this schema:
+
+- **"On-demand" updates**: with this strategy, you and your team have agreed that updates can be made to the schema as needed. However, in order to avoid migration failures due to inconsistencies between the data and schema, there is agreement in the team that any new fields added are explicitly defined as optional.
+
+  In our scenario above, you can go ahead and add the new `billingAddress` field to the `User` model in your Prisma schema, and define it as optional:
+
+  ```prisma file=prisma/schema.prisma highlight=10;add
+  model Product {
+    id     String  @id @default(auto()) @map("_id") @db.ObjectId
+    name   String  @unique
+  }
+
+  model Order {
+    id              String   @id @default(auto()) @map("_id") @db.ObjectId
+    product         Product  @relation(fields: [productId], references: [id])
+    shippingAddress Address
+    billingAddress  Address?
+    productId       String   @db.ObjectId
+  }
+
+  type Address {
+    street String @id @default(auto()) @map("_id") @db.ObjectId
+    city   String
+    zip    String
+  }
+  ```
+
+  Then regenerate your Prisma Client using the `npx prisma generate` command. Next, update your application to reflect the new field, and redeploy your app. The records in the database will be updated as the application's users begin to enter their billing address in the new field.
+
+- **"No breaking changes" updates**: this strategy builds on the first one, with further consensus amongst your team that you don't rename or delete fields, only add new fields, and always define the new fields as optional. This policy can be reenforced by adding checks in the CI/CD process to verify that there are no backwards-incompatible changes to the schema.
```

Comment:
Isn't this what the first example also does?

Reply:
Yes, it is basically same, but the strategy is slightly diff in that it includes the sue of CI/CD checks.. seems nice to call that out as an additional approach. It's a little more formal than the "on demand" strategy.

Reply:
For me this is the same as the first, as the first one actually only works if all the things described here are the case.

How would a CI check for such "backwards-incompatible changes" look like?

Reply:
We are presenting these three options as "strategies"... as approaches that our users can think about and consider if they want to use. These are not instructions or processes, but rather strategies they might want to employ. SO I think it is very valuable to keep this second one; maybe readers wouldn't even consider CI/CD checks unless they read this. And we can't tell them exactly how to incorporate any of the checks into their CI/CD build, but I could imagine that they read this and think "Oh, yeah, we are a big team, we should totally add a rule in our build that checks for new fields that have not been defined as optional, and stop the build if any are found."

Reply:
Then I do not understand why we gave strategy 2 the name that describes what we do in strategy 1, but add "use CI/CD" to the description (without adding it to the name of the strategy). 

To me this was super confusing and I could not differentiate strategy 1 and 2 and the names to not related directly to what they describe and how they are different from each other.

Reply:
Hmmm maybe the names need to be tweaked. I do think it is worth calling out all three, though.

Perhaps this or something like it:
#1 **On-demand, passive updates** (I use the word passive because the data is passively updated as the app is used, but that might be a misleading word for other reasons)
#2 **CI/CD-enforced** (and still explain that this strategy builds on #3, but if enforced/verified)
#3 **All-at-once**

We have a little more time... will get it cleaned up/decided before tomorrow's release, and can even iterate afterwards. @keerlu of course feel free to edit as you want; I am happy with you and @matthewmueller being the final decision makers on this. :-).

Reply:
I like the term passive updates.

I see #2 as more of an organizational rule or process. For example, the MongoDB Atlas team themselves applies this rule when administering their own MongoDB database.

I see it as different, but not sure the best way to describe it.

Reply:
Will revisit improving the wording after release